deeplearning4j/deeplearning4j-cuda/src/main/java/org/deeplearning4j/nn/layers/dropout/CudnnDropoutHelper.java

```diff
@@ -0,0 +1,147 @@
+
+        public CudnnDropoutContext() {
+            createHandles();
+            deallocator(new Deallocator(this));
+        }
+
+        public CudnnDropoutContext(CudnnDropoutContext c) {
+            super(c);
+            xTensorDesc = new cudnn.cudnnTensorStruct(c.xTensorDesc);
+            dxTensorDesc = new cudnn.cudnnTensorStruct(c.dxTensorDesc);
+            yTensorDesc = new cudnn.cudnnTensorStruct(c.yTensorDesc);
+            dyTensorDesc = new cudnn.cudnnTensorStruct(c.dyTensorDesc);
+            dropoutDesc = new cudnn.cudnnDropoutStruct(c.dropoutDesc);
+        }
+
+        @Override
+        protected void createHandles() {
+            super.createHandles();
+            checkCudnn(cudnnCreateTensorDescriptor(xTensorDesc));
+            checkCudnn(cudnnCreateTensorDescriptor(dxTensorDesc));
+            checkCudnn(cudnnCreateTensorDescriptor(yTensorDesc));
+            checkCudnn(cudnnCreateTensorDescriptor(dyTensorDesc));
+            checkCudnn(cudnnCreateDropoutDescriptor(dropoutDesc));
+        }
+
+        @Override
+        protected void destroyHandles() {
+            checkCudnn(cudnnDestroyTensorDescriptor(xTensorDesc));
+            checkCudnn(cudnnDestroyTensorDescriptor(dxTensorDesc));
+            checkCudnn(cudnnDestroyTensorDescriptor(yTensorDesc));
+            checkCudnn(cudnnDestroyTensorDescriptor(dyTensorDesc));
+            checkCudnn(cudnnDestroyDropoutDescriptor(dropoutDesc));
+            super.destroyHandles();
+        }
+    }
+
+    private CudnnDropoutContext cudnnContext = new CudnnDropoutContext();
+    private Pointer statesPtr;
+    private long stateSizeBytes;
+    private long seed;
+
+
+    @Override
+    public void applyDropout(INDArray input, INDArray resultArray, double dropoutInputRetainProb) {
+        float p = (float)(1.0 - dropoutInputRetainProb);    //CuDNN uses p = probability of setting to 0
+
+        checkCudnn(cudnnSetTensorNdDescriptor(cudnnContext.xTensorDesc, dataType, input.rank(),
+                //TODO int cast
+                ArrayUtil.toInts(input.shape()),
+                ArrayUtil.toInts(input.stride())));
+        checkCudnn(cudnnSetTensorNdDescriptor(cudnnContext.yTensorDesc, dataType, resultArray.rank(),
+                //TODO int cast
+                ArrayUtil.toInts(resultArray.shape()),
+                ArrayUtil.toInts(resultArray.stride())));
+
+        //Reserve space
+        SizeTPointer stateSizeBytesPtr = new SizeTPointer();        //TODO correct???
+        stateSizeBytes = stateSizeBytesPtr.get();
+        checkCudnn(cudnnDropoutGetReserveSpaceSize(cudnnContext.xTensorDesc, stateSizeBytesPtr));
+
+        //Dropout descriptor:
+        if(statesPtr == null || statesPtr.capacity() != stateSizeBytes){
+            //TODO: Is this correct? "Pointer to user-allocated GPU memory that will hold random number generator states."
+            statesPtr = new BytePointer(stateSizeBytes);
+        }
```

Comment:
It probably needs memory allocated on the device. Let's try to use `DataCache` available in the parent and as used by `CudnnConvolutionHelper`, `CudnnBatchNormalizationHelper`, and `CudnnLSTMHelper`.

Reply:
We should also call `states.deallocate()` just before `new DataCache()` to free up precious device memory right away.